File: Python/array/find_all_numbers_disappeared_in_an_array.py
from typing import List


def findDisappearedNumbers(nums: List[int]) -> List[int]:
    n = len(nums)
    for num in nums:
        num = num % n
        nums[num - 1] += n

    result = []
    for i in range(n):
        if nums[i] <= n:
            result.append(i + 1)

    return result


if __name__ == "__main__":
    print(findDisappearedNumbers([4, 3, 2, 7, 8, 2, 3, 1]))

File: core/tmp/Java/src/com/dbc/algorithm/Array/FindAllNumbersDisappearedInAnArray.java
package com.dbc.algorithm.Array;

import java.util.ArrayList;
import java.util.List;

public class FindAllNumbersDisappearedInAnArray {
    public static List<Integer> findDisappearedNumbers(int[] nums) {
        for (int num : nums) {
            num = (num - 1) % nums.length;
            nums[num] += nums.length;
        }

        List<Integer> result = new ArrayList<>();
        for (int i = 0; i < nums.length; i++) {
            if (nums[i] <= nums.length) {
                result.add(i + 1);
            }
        }
        return result;
    }
}
